### PR TITLE
VNF delays

### PR DESCRIPTION
**Pull request by stefanbschneider**
Specify, read, and consider VNF delays in total delay

With this PR:
* VNF delays can be specified in a template (optional; default = 0 delay)
* These are read and considered by B-JointSP when creating the components and when calculating the total delay in the objective value of a placement
* The vnf delay is separately saved in the written results

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+# Modifications by Haydar Qarawlus and Adnan
```
Comment (stefanbschneider) [inline setup.py]:
```suggestion
```

File: src/bjointsp/parameters/templates/fw1chain.yaml [hunk 2]
```diff
@@ -9,22 +9,31 @@ vnfs:
     inputs_bwd: 0
     outputs_fwd: 1
     outputs_bwd: 0
-    cpu: [0]
```
Comment (stefanbschneider) [inline src/bjointsp/parameters/templates/fw1chain.yaml]: Please avoid changes that are unrelated to VNF processing delays. Setting CPU requirements for the user "vnf" dosn't really make sense

File: src/bjointsp/template/component.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
 class Component:
-    def __init__(self, name, type, stateful, inputs, outputs, cpu, mem, dr, config=None):
+    def __init__(self, name, type, stateful, inputs, outputs, cpu, mem, dr, vnf_delay, config=None):
```
Comment (stefanbschneider) [inline src/bjointsp/template/component.py]: `vnf_delay` should be optional. `vnf_delay=None`; check if it's assigned otherwise set to 0 (and write debug logging comment).

File: src/bjointsp/parameters/templates/fw1chain.yaml [hunk 4]
```diff
@@ -9,22 +9,31 @@ vnfs:
...
+    cpu: [1]
     mem: [0]
+    vnf_delay: 50
```
Comment (stefanbschneider) [inline src/bjointsp/parameters/templates/fw1chain.yaml]: Similarly, I don't think setting "vnf_delay" for the user "vnf" makes sense.

 In fact, maybe you could add a check in the `reader.py` that checks if all resource requirements and vnf_processing delays for user/source vnfs are set to 0. If this is not the case, log a warning message (but don't throw an exception)

File: src/bjointsp/read_write/writer.py [hunk 5]
```diff
@@ -90,9 +90,9 @@ def save_heuristic_variables(result, changed_instances, instances, edges, nodes,
             result["placement"]["flows"].append(flow)
         for path in e.paths:
             # record edge delay: all flows take the same (shortest) path => take path delay
-            delay = {"src": e.arc.source.name, "dest": e.arc.dest.name, "src_node": e.source.location, "dest_node": e.dest.location, "delay": sp.path_delay(links, path)}
```
Comment (stefanbschneider) [inline src/bjointsp/read_write/writer.py]: Could you please adjust this a little bit to have a list of `path_delays` that contains the original flow/link delays between interconnected VNFs and a separate block `vnf_delays` that contains the vnf processing delays. Mixing the two makes the results file harder to interpret.

File: src/bjointsp/read_write/writer.py [hunk 6]
```diff
@@ -90,9 +90,9 @@ def save_heuristic_variables(result, changed_instances, instances, edges, nodes,
...
+            delay = {"src": e.arc.source.name, "dest": e.arc.dest.name, "src_node": e.source.location, "dest_node": e.dest.location, "path_delay": sp.path_delay(links, path), "src_vnf_delay": e.arc.source.vnf_delay, "dest_vnf_delay": e.arc.dest.vnf_delay }
             result["metrics"]["delays"].append(delay)
-            result["metrics"]["total_delay"] += sp.path_delay(links, path)
+            result["metrics"]["total_delay"] += sp.path_delay(links, path)+e.arc.source.vnf_delay+e.arc.dest.vnf_delay
```
Comment (stefanbschneider) [inline src/bjointsp/read_write/writer.py]: This is wrong because it will add the vnf processing delays multiple times. Eg, in a chain of A->B->C, the VNF delay of B is added as dest_vnf in A->B and as src vnf in B->C.

Please just add vnf processing delays in a separate loop and don't mix it with the link delays. that's cleaner anyways

File: src/bjointsp/template/component.py (hunk 3 above)
Comment (stefanbschneider) [reply src/bjointsp/template/component.py]: In your example, the total delay should be 105, but it is calculated (incorrectly) as 140 (because 35 is added twice).

File: src/bjointsp/template/component.py [hunk 7]
```diff
@@ -19,11 +19,14 @@ def __init__(self, name, type, stateful, inputs, outputs, cpu, mem, dr, config=N
...
         self.cpu = cpu      # function of forward and backward ingoing data rates
         self.mem = mem
+        self.vnf_delay = vnf_delay
         self.dr = dr[0]
         self.dr_back = dr[1]
         self.config = config		# config used by external apps/MANOs (describes image, ports, ...)
 
         total_inputs = self.inputs + self.inputs_back
+
+        # What is this used for? For every input we need a total of 1 CPU and 1 Mem?
```
Comment (stefanbschneider) [inline src/bjointsp/template/component.py]: As discussed last time, the number of CPU/memory coefficients depends on the number of inputs. If we have a VNF with 2 inputs, we'll need 3 CPU parameters: a*lambda1+b*lambda2+c

This simply checks that the number of CPU parameters matches the number of inputs. It does not mean that we have to have a total of 1 CPU or 1 Memory.

File: src/bjointsp/parameters/templates/fw1chain.yaml (hunk 2 above)
Comment (qarawlus) [reply src/bjointsp/parameters/templates/fw1chain.yaml]: I copied this back from fw2chain.yaml after I made some changes and was unsure of what it was like before. In that file and fw3chaib.yaml the user vnf has both cpu and mem requirements.

**Comment by stefanbschneider** [comment]
Thanks for the work, but there are a number of things, you should fix before I can merge the pull request (also see in-line comments and suggestions in the various files above):

- [ ] why did you change the function to sort templates according to their weight? The idea there was to embed templates first that require lots of resources. I don't think including vnf_delay here makes sense...
- [ ] remove meaningless/debug comments. committed comments should be to help understand your code only. please check all files again
- [ ] fix the calculation of the total delay in the results file (see inline comment).
- [ ] when recording the individual delays in the results file, make separate blocks for `path_delays` and `vnf_delays`. don't mix them
- [ ] you should also conisder the vnf processing delay in the objective function, where the quality of a potential placement is checked. here: https://github.com/qarawlus/B-JointSP/blob/dev/src/bjointsp/heuristic/control.py#L72

File: src/bjointsp/template/component.py (hunk 3 above)
Comment (qarawlus) [reply src/bjointsp/template/component.py]: The default is already set in the reader file, so if it doesn't find a vnf_delay in the yaml file it will default to zero. I will add a default in the function definition as well.

File: src/bjointsp/parameters/templates/fw1chain.yaml (hunk 2 above)
Comment (stefanbschneider) [reply src/bjointsp/parameters/templates/fw1chain.yaml]: No problem. You can always easily check the file diffs in the pull request (here). It'll show you exactly what you changed

**Comment by qarawlus** [comment]
Will work on all the above and fix/adjust as soon as possible. Thanks a lot for the feedback.

**Comment by stefanbschneider** [comment]
Perfect, thanks! :)

File: src/bjointsp/read_write/writer.py [hunk 8]
```diff
@@ -101,6 +103,12 @@ def save_heuristic_variables(result, changed_instances, instances, edges, nodes,
                     consumed_dr[(path[i], path[i+1])] += e.flow_dr() / len(e.paths)
                     link = {"arc": str(e.arc), "edge_src": e.source.location, "edge_dst": e.dest.location, "link_src": path[i], "link_dst": path[i+1]}
                     result["placement"]["links"].append(link)
+        for path in e.paths:
+            # record vnf delays
+            vnf_delay = {"dest_vnf": e.arc.dest.name, "dest_vnf_delay": e.arc.dest.vnf_delay }
+            result["metrics"]["vnf_delays"].append(vnf_delay)
+            result["metrics"]["total_vnf_delay"] += e.arc.dest.vnf_delay
+
```
Comment (stefanbschneider) [inline src/bjointsp/read_write/writer.py]: Nice, the `total_path_delay` and `total_vnf_delay` look good. To minor suggestions:

* In the `vnf_delays` section, simply call the fields `vnf` and `vnf_delay` instead of `dest_vnf` and `dest_vnf_delay`. I think it's a bit clearer when looking at the result file.
* Could you add a third field `total_delay = total_path_delay + total_vnf_delay` for convenience?

File: src/bjointsp/template/template.py [hunk 9]
```diff
@@ -49,11 +49,12 @@ def weight(self, src_dr):
         # iterate over components in topological order, determine outgoing dr for each and store it in out_dr
         # first, consider only forward direction then only backward
         # also determine cpu/mem consumption per component and direction and increase total_cpu/mem accordingly
-        total_cpu, total_mem = 0, 0
+        total_cpu, total_mem = 0, 0,
```
Comment (stefanbschneider) [inline src/bjointsp/template/template.py]: Please remove the comma. Probably doesn't break anything, but just to be on the safe side.

File: src/bjointsp/heuristic/control.py [hunk 10]
```diff
@@ -112,7 +116,7 @@ def objective_value(overlays, print_info=False):
 
     # minimize total delay
     elif obj == objective.DELAY:
-        value = total_delay
+        value = total_delay+vnf_delays
```
Comment (stefanbschneider) [inline src/bjointsp/heuristic/control.py]: One more, important thing, I just noticed. There are 5 possible objectives for B-JointSP: Combined (default), over-subscription only, number of changed instances only, resources only, delay only.

You should add the `vnf_delay` to the total delay earlier, right after the place where it's calculated (currently line 79 or so). Currently, the vnf_delay is only added if the delay minimization objective is selected, but for example not for the default combined objective.

**Comment by qarawlus** [comment]
Thanks for the comments. 
I have made the required adjustments to the code.
